Спринт_09 первая тема

Williamwhite25 opened the pull request:
-Создан соответствующий файл ProfileViewController
-Добавлен новый контроллер в Main.storyboard
-Установлен класс ProfileViewController новому контроллеру -Добавлено в контроллер UIImageView, три UILabel и один UIButton -Расставлены констрейнты нужным образом
-Выгружены из Figma картинка для профиля и картинка для кнопки логаута, подставлены  в нужные элементы в Storyboard -Созданы соответствующие @IBOutlet'ы и @IBAction

Williamwhite25 committed:
-Создан соответствующий файл ProfileViewController
-Добавлен новый контроллер в Main.storyboard
-Установлен класс ProfileViewController новому контроллеру
-Добавлено в контроллер UIImageView, три UILabel и один UIButton
-Расставлены констрейнты нужным образом
-Выгружены из Figma картинка для профиля и картинка для кнопки логаута, подставлены  в нужные элементы в Storyboard
-Созданы соответствующие @IBOutlet'ы и @IBAction

Williamwhite25 committed:
Сохраняю для себя имплементацию через Navigation Controller

Williamwhite25 committed:
Sprint_09 готов

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,15 @@
+    
+    @IBOutlet var cellImage: UIImageView!
+    @IBOutlet var likeButton: UIButton!
+    @IBOutlet var dateLabel: UILabel!
```
В твоём классе ImagesListCell, свойства @IBOutlet ссылаются на элементы интерфейса (например, UIImageView, UIButton, UILabel), которые создаются и управляются Storyboard или XIB-файлом.

Когда ты подключаешь UI-элемент через @IBOutlet, он уже удерживается в памяти иерархией UIView, которая в свою очередь удерживается системой UIViewController/UITableViewCell. Если ты не укажешь weak, то создаётся сильная ссылка (strong reference), и это может привести к retain cycle — особенно в случае, если один объект косвенно удерживает другой, и они никогда не освобождаются из памяти.

Поэтому в большинстве случаев Apple рекомендует использовать weak для @IBOutlet, чтобы избежать таких проблем.

💡 Как правильно:
```

@IBOutlet weak var cellImage: UIImageView!
@IBOutlet weak var likeButton: UIButton!
@IBOutlet weak var dateLabel: UILabel!

```

📌 Кратко:

Используем weak для @IBOutlet, чтобы избежать утечек памяти и циклов удержания (retain cycles), особенно в интерфейсе, который управляется storyboard'ом или XIB'ом.

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+import UIKit
+
+class ImagesListViewController: UIViewController {
```
Ты создаёшь класс, который можно наследовать. Но если ты не планируешь, чтобы кто-то создавал подклассы от ImagesListViewController, то лучше сразу объявить его как final:
```
final class ImagesListViewController: UIViewController {

```

📌 Зачем?

Повышение производительности: Компилятор знает, что класс final, и может делать более эффективную оптимизацию (например, вызовы методов без виртуальной таблицы).

Безопасность архитектуры: Ты явно говоришь: «этот класс не предназначен для наследования». Это помогает другим разработчикам (и тебе самому в будущем) не делать ошибок в архитектуре.

Прозрачность намерений: Это улучшает читаемость и поддержку кода — видно, что поведение контроллера не должно переопределяться.

IgorChumikov replied:
final

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+class ImagesListViewController: UIViewController {
+    private let showSingleImageSegueIdentifier = "ShowSingleImage"
+    
+    @IBOutlet private var tableView: UITableView!
```
weak

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+import UIKit
+
+class ImagesListViewController: UIViewController {
+    private let showSingleImageSegueIdentifier = "ShowSingleImage"
```
🔍 Проблемы:

Название переменной слишком длинное и «шумное» — читается тяжело.

Дублирует саму строку "ShowSingleImage", что снижает читабельность.

В коде будет длинно выглядеть: performSegue(withIdentifier: showSingleImageSegueIdentifier, ...)

✅ Как лучше:
```
private let segueID = "ShowSingleImage"

```

🔧 Почему так лучше:

Коротко и понятно, особенно если этот segue — единственный или основной в контроллере.
Название всё ещё ясно указывает на назначение — segue ID.
Код становится короче и легче читается:

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+    
+    @IBOutlet private var tableView: UITableView!
+    
+    private let photosName: [String] = Array(0..<20).map { "\($0)" }
```
✅ Улучшенный вариант:

```
private let photoNames = (0..<20).map(String.init)
```

🔍 Почему так лучше:

Без лишнего шума:
Убрал Array(...), так как map уже возвращает массив.

Использование стандартных API:
String.init — это более лаконичный и читаемый способ конвертировать числа в строки.

Правильное имя переменной:
photoNames (множественное число) — корректнее с точки зрения английского, чем photosName.

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        tableView.rowHeight = 200
```
можно удалить строку:  
```
tableView.rowHeight = 200
```
🔍 Причина:

В коде у тебя уже реализован метод:
```
func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat
```
Этот метод делегата UITableViewDelegate переопределяет поведение rowHeight, вычисляя высоту каждой ячейки динамически, на основе размера изображения.

➡️ А значит: tableView.rowHeight = 200 становится бесполезным, потому что она не используется — конкретная высота всё равно берётся из heightForRowAt.

IgorChumikov commented on ImageFeed/ViewControllers/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,116 @@
+        tableView.contentInset = UIEdgeInsets(top: 12, left: 0, bottom: 12, right: 0)
+    }
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
✅ Улучшенный вариант:
```

override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    guard segue.identifier == showSingleImageSegueIdentifier else {
        super.prepare(for: segue, sender: sender)
        return
    }

    guard
        let viewController = segue.destination as? SingleImageViewController,
        let indexPath = sender as? IndexPath
    else {
        assertionFailure("Invalid segue destination or sender type")
        return
    }

    viewController.image = UIImage(named: photosName[indexPath.row])
}

```

🔍 Что улучшено и почему:

Упрощён контроль segue:
Вместо вложенного if, теперь используется guard — читается легче и структурно чище.

Унифицирована проверка на segue.identifier:
Сразу возвращаемся в else, если не тот segue — меньше вложенности, меньше когнитивной нагрузки.

Ясное сообщение в assertionFailure:
Теперь ошибка говорит о двойной причине: либо segue.destination не тот, либо sender не IndexPath.

Удалена промежуточная переменная image:
Она использовалась один раз — можно спокойно сократить строку.

IgorChumikov commented on ImageFeed/ViewControllers/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,120 @@
+//    @IBAction func didTapLogoutButton() {
+//    }
+//    
+//}
```
Лишний код лучше удалить если он не используются

IgorChumikov commented on ImageFeed/ViewControllers/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,120 @@
+
+
+
+//Я правильно понял, что @IBOutlet надо удалить? Я на всякий сслучай оставил. Напишите ответ после проверки, спасибо!
```
Да, ты всё понял правильно — если ты полностью перешёл на программный UI, @IBOutlet тебе больше не нужны. Можешь смело удалять закомментированный код с ними. Молодец, что подстраховался, но сейчас уже можно очистить файл 🧹

IgorChumikov commented on ImageFeed/ViewControllers/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+    
+    @IBOutlet private var imageView: UIImageView!
+    @IBOutlet private var scrollView: UIScrollView!
```
waek

IgorChumikov commented on ImageFeed/ViewControllers/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,120 @@
+//Я правильно понял, что @IBOutlet надо удалить? Я на всякий сслучай оставил. Напишите ответ после проверки, спасибо!
+
+import UIKit
+
```
💡 Что такое // MARK:?

Это специальная пометка-комментарий, которую Xcode понимает и отображает в виде разделов в выпадающем списке методов (в навигаторе и списке символов в редакторе). Она не влияет на выполнение программы, но делает код аккуратным и логически разделённым.

✅ Вот как можно красиво оформить ProfileViewController:
```

import UIKit

final class ProfileViewController: UIViewController {
    
    // MARK: - UI Elements
    
    private var avatarImageView: UIImageView!
    private var nameLabel: UILabel!
    private var loginNameLabel: UILabel!
    private var descriptionLabel: UILabel!
    private var logoutButton: UIButton!
    
    // MARK: - Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupAvatarImageView()
        setupNameLabel()
        setupLoginNameLabel()
        setupDescriptionLabel()
        setupLogoutButton()
    }
    
    // MARK: - Setup UI
    
    private func setupAvatarImageView() {
        // ... (код создания и добавления аватарки)
    }

    private func setupNameLabel() {
        // ... (код для nameLabel)
    }

    private func setupLoginNameLabel() {
        // ... (код для loginNameLabel)
    }

    private func setupDescriptionLabel() {
        // ... (код для descriptionLabel)
    }

    private func setupLogoutButton() {
        // ... (код для logoutButton)
    }

    // MARK: - Actions

    @objc
    private func didTapLogoutButton() {
        // Обработка выхода
    }
}

```
📌 Объяснение

// MARK: помогает быстро понять, где что находится в классе: где свойства, где методы, где действия. В больших проектах это реально спасает, особенно когда много логики и людей в команде.

Williamwhite25 committed:
Добавлены:
-  weak к @IBOutlet,
- final к классам
-МАРК

Изменено:
- Название переменной с private let showSingleImageSegueIdentifier на private let segueID
-  private let photosName: [String] = Array(0..<20).map { "\($0)" } на private let photoNames = (0..<20).map(String.ini

Williamwhite25 committed:
Чек

Williamwhite25 committed:
1
